feat: Add FP4 (E2M1) KV Cache Support with Quantization Utilities for MLA

JackChuang opened the pull request:
## Summary
This PR introduces FP4 (E2M1) support for Multi-Head Latent Attention (MLA) KV cache in SGLang, enabling low-precision caching to reduce memory usage and improve inference efficiency. It integrates FP4 quantization utilities, Triton kernels, and unit tests while remaining backward compatible with FP16/FP8. See #10083, points 1-1, for more context.

Co-authored-by: @yicwang  Yichen Wang <yichen.wang@bytedance.com> 

## Usage
Added `--kv-cache-dtype=fp4_e2m1` option.
``` 
$ python3 -m sglang.launch_server --kv-cache-dtype fp4_e2m1 ... 
```

## Key Changes
1. **Server Argument Extension**
   - Added `--kv-cache-dtype=fp4_e2m1` option.

2. **FP4 Quantization Utility**
   - Added `KVFP4QuantizeUtil` with `batched_quantize` and `batched_dequantize` methods for block-wise (16) processing of [M, N, K] tensors.

3. **Core KV Cache Integration**
   - Updated `ModelRunner` and `MLATokenToKVPool` to support FP4 KV cache.
   - Added `kv_scale_buffer` for FP4 scaling.
   - Implemented Triton kernel `set_mla_kv_scale_buffer_kernel` for efficient nope+rope tensor handling.
   - Maintains backward compatibility with FP16/FP8 KV cache.

4. **Unit Test & Benchmark**
   - Tests validate FP4 correctness via MSE, MAE, PSNR, and Relative Error.
   - Benchmarks FP4 vs FP8 performance on GPU.

## Impact
- Enables low-precision FP4 KV cache for MLA.
- Reduces GPU memory usage while maintaining accuracy.
- Provides measurable speedup over FP8 quantization.
- Backward compatible with existing FP16/FP8 workflows.

## Accuracy Tests
The results show that on simpler datasets, the accuracy is nearly lossless compared to the baseline. On more challenging datasets, there is some accuracy degradation, but it remains within an acceptable range.
``` 
| Model | Dataset | Metric | Subset. | Num | Score | Cat.0 |


| DeepSeek-R1-0528-FP4-KV16 | gsm8k | mean_acc | main | 6595 | 0.9157 | default |
| DeepSeek-R1-0528-FP4-KV8_e4m3 | gsm8k | mean_acc | main | 6595 | 0.9154 | default |
| DeepSeek-R1-0528-FP4-KV4_fp4_e2m1 | gsm8k | mean_acc | main | 6595 | 0.9124 | default |

| DeepSeek-R1-0528-FP4-KV16 | aime25 | mean_acc | OVERALL | 150 | 0.5067 | - |
| DeepSeek-R1-0528-FP4-KV8_e4m3 | aime25 | mean_acc | OVERALL | 150 | 0.4934 | - |
| DeepSeek-R1-0528-FP4-KV4_fp4_e2m1 | aime25 | mean_acc | OVERALL | 150 | 0.4 | - |


| DeepSeek-R1-0528-FP4-KV16 | gpqa_diamond | mean_acc | default | 990 | 0.7707 | default |
| DeepSeek-R1-0528-FP4-KV8_e4m3 | gpqa_diamond | mean_acc | default | 990 | 0.7697 | default |
| DeepSeek-R1-0528-FP4-KV4_fp4_e2m1 | gpqa_diamond | mean_acc | default | 990 | 0.7273 | default |

```


## Performance Results
Tested on B200. Server is running with --model DeepSeek-R1-0528-FP4 --tp-size 4 --moe-runner-backend flashinfer_trtllm --disable-radix-cache. Client is running with with --goodput ttft:5000 tpot:50 --random-input-len 3500 --random-output-len 1500 --max-concurrency 50 and --num-prompts 100.

Baseline has the best performance as the trtllm kernel accepts the BF16 KVCache directly, so no dequantization overhead at all. 
```
(Default) BF16 KVCache

============ Serving Benchmark Result ============
Successful requests: 100
Benchmark duration (s): 116.44
Total input tokens: 349051
Total generated tokens: 150000
Request throughput (req/s): 0.86
Request goodput (req/s): 0.46
Output token throughput (tok/s): 1288.17
Total Token throughput (tok/s): 4285.73
---------------Time to First Token----------------
Mean TTFT (ms): 4853.44
Median TTFT (ms): 4784.58
P50 TTFT (ms): 4784.58
P90 TTFT (ms): 7828.38
P95 TTFT (ms): 9386.72
P99 TTFT (ms): 9616.72
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms): 35.05
Median TPOT (ms): 35.11
P50 TPOT (ms): 35.11
P90 TPOT (ms): 37.41
P95 TPOT (ms): 37.88
P99 TPOT (ms): 37.98
---------------Inter-token Latency----------------
Mean ITL (ms): 35.05
Median ITL (ms): 32.00
P50 ITL (ms): 32.00
P90 ITL (ms): 33.60
P95 ITL (ms): 33.91
P99 ITL (ms): 34.32
==================================================
```


When adding --kv-cache-dtype fp8_e4m3, the performance dropped significantly as KVCache is quanted at writing and dequanted at reading for attentions.
```
--kv-cache-dtype fp8_e4m3

============ Serving Benchmark Result ============
Successful requests: 100
Benchmark duration (s): 474.62
Total input tokens: 349051
Total generated tokens: 150000
Request throughput (req/s): 0.21
Request goodput (req/s): 0.00
Output token throughput (tok/s): 316.05
Total Token throughput (tok/s): 1051.48
---------------Time to First Token----------------
Mean TTFT (ms): 6093.23
Median TTFT (ms): 6057.21
P50 TTFT (ms): 6057.21
P90 TTFT (ms): 9279.85
P95 TTFT (ms): 10945.23
P99 TTFT (ms): 11347.56
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms): 153.69
Median TPOT (ms): 153.72
P50 TPOT (ms): 153.72
P90 TPOT (ms): 156.34
P95 TPOT (ms): 156.80
P99 TPOT (ms): 157.09
---------------Inter-token Latency----------------
Mean ITL (ms): 153.69
Median ITL (ms): 150.40
P50 ITL (ms): 150.40
P90 ITL (ms): 151.99
P95 ITL (ms): 152.28
P99 ITL (ms): 152.68
==================================================
```

With --kv-cache-dtype fp4_e2m1, the throughput is 17.8% higher compared to fp8_e4m3. Mostly due to quant/dequant to/from FP4 is faster than FP8.
```
--kv-cache-dtype fp4_e2m1

============ Serving Benchmark Result ============
Successful requests: 100
Benchmark duration (s): 402.92
Total input tokens: 349051
Total generated tokens: 150000
Request throughput (req/s): 0.25
Request goodput (req/s): 0.00
Output token throughput (tok/s): 372.28
Total Token throughput (tok/s): 1238.58
---------------Time to First Token----------------
Mean TTFT (ms): 5611.71
Median TTFT (ms): 5675.33
P50 TTFT (ms): 5675.33
P90 TTFT (ms): 9137.97
P95 TTFT (ms): 9952.08
P99 TTFT (ms): 10323.24
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms): 130.10
Median TPOT (ms): 129.87
P50 TPOT (ms): 129.87
P90 TPOT (ms): 133.07
P95 TPOT (ms): 133.56
P99 TPOT (ms): 133.66
---------------Inter-token Latency----------------
Mean ITL (ms): 130.10
Median ITL (ms): 126.83
P50 ITL (ms): 126.83
P90 ITL (ms): 128.53
P95 ITL (ms): 128.91
P99 ITL (ms): 129.45
==================================================
```

## Future Work
We plan to support Multi-Head Attention (MHA) next.


## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

JackChuang committed:
Add FP4 support for KV cache dtype in ServerArgs

Extend the `--kv-cache-dtype` argument in ServerArgs to support "fp4_e2m1".

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Yichen Wang <yichen.wang@bytedance.com>

JackChuang committed:
Add KVFP4 quantization utility for KV cache

- Introduce `KVFP4QuantizeUtil` for FP4 (E2M1) quantization and dequantization.
- Provides `batched_quantize` and `batched_dequantize` methods for block-wise (16) processing of [M, N, K] tensors.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Yichen Wang <yichen.wang@bytedance.com>

JackChuang committed:
Add unit test and benchmark for KVFP4 quantization

- Introduce `test_kvfp4_quant_dequant.py` to validate correctness and performance of KVFP4 quantization.
- Provides metrics calculation (MSE, MAE, PSNR, Relative Error) to compare original and dequantized tensors.
- Benchmarks KVFP4 vs FP8 quant/dequant performance on GPU with large tensors.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Yichen Wang <yichen.wang@bytedance.com>

JackChuang committed:
Add FP4 KV cache support with scale buffer and Triton kernels for MLA

Core change enabling low-precision FP4 KV caching for MLA,
improving inference efficiency while keeping existing workflows intact.

- Introduce FP4 KV cache support in MLATokenToKVPool for reduced memory usage.
- Add kv_scale_buffer to store FP4 scaling factors and updated allocation logic.
- Implement Triton kernel to combine nope + rope tensors and write to KV + scale buffers.
- Modify ModelRunner to account for FP4 buffer sizing and dtype.
- Maintains backward compatibility with FP16/FP8 KV cache.

Also, move Triton kernels from mem_cache/memory_pool.py to srt/mem_cache/utils.py.

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>
Co-authored-by: Yichen Wang <yichen.wang@bytedance.com>

AniZpZ commented on python/sglang/srt/server_args.py:
```diff
@@ -987,8 +987,8 @@ def add_cli_args(parser: argparse.ArgumentParser):
-            choices=["auto", "fp8_e5m2", "fp8_e4m3"],
-            help='Data type for kv cache storage. "auto" will use model data type. "fp8_e5m2" and "fp8_e4m3" is supported for CUDA 11.8+.',
+            choices=["auto", "fp8_e5m2", "fp8_e4m3", "fp4_e2m1"],
+            help='Data type for kv cache storage. "auto" will use model data type. "fp8_e5m2" and "fp8_e4m3" is supported for CUDA 11.8+. "fp4_e2m1" is supported for CUDA 12.8+',
```
i think it might be better to clarify that only nvfp4 e2m1 but not mxfp4 e4m1 is supported

JackChuang replied:
Will do. Thanks!

AniZpZ commented:
Are there any more tests to evaluate accuracy drop in long-context scenarios?

AniZpZ commented on python/sglang/srt/model_executor/model_runner.py:
```diff
@@ -1065,6 +1065,21 @@ def profile_max_num_token(self, total_gpu_memory: int):
                 * torch._utils._element_size(self.kv_cache_dtype)
             )
+
+            if self.kv_cache_dtype == torch.float4_e2m1fn_x2:
```
check cuda

AniZpZ commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -753,15 +820,41 @@ def __init__(
-                    )
-                    for _ in range(layer_num)
-                ]
+                if self.dtype != torch.float4_e2m1fn_x2:
```
check cuda

JackChuang commented:
> Are there any more tests to evaluate accuracy drop in long-context scenarios?

Do you have a specific dataset result you’d like to see?  We can collect the data.

AniZpZ reviewed (approved):
LGTM
but i think we'd better note that FP4 (E2M1) KV Cache might lead to a accuracy drop

cc @JackChuang  @zhyncs

JackChuang commented:
> LGTM but i think we'd better note that FP4 (E2M1) KV Cache might lead to a accuracy drop

@AniZpZ Sounds good. Where do you think would be the best place for us to note this?

JackChuang commented:
> https://github.com/sgl-project/sglang/pull/10078/files#diff-65c6ac2c41977f68e460f18e35053b97089631f88a9958b0796343fccee78a67R1205
> 
> Add a log here to note the accuracy issue. Quantization docs will be revised in a few days

Done and pushed. Added a logger.warning().

JackChuang commented:
Hi @AniZpZ @b8zhong @zhyncs, I’ve checked the CI failures again, and they don’t seem related to my changes.
If there are no further comments, would it be possible for a maintainer to help merge this PR? Thanks!

Fridge003 commented on python/sglang/srt/layers/quantization/kvfp4_tensor.py:
```diff
@@ -0,0 +1,113 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Can you paste the link to the source of these codes?

Fridge003 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -1017,6 +1017,73 @@ def set_mla_kv_buffer_triton(
     )
 
 
+@triton.jit
```
Maybe we can open a new utils.py file for storing these triton kernels?

Fridge003 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -1017,6 +1017,73 @@ def set_mla_kv_buffer_triton(
+    )
+
+
 class MLATokenToKVPool(KVCache):
```
Can we make FP4 MLA Pool a derived class of `MLATokenToKVPool`? So that we don't need these if-else conditions.

Fridge003 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -1017,6 +1017,73 @@ def set_mla_kv_buffer_triton(
+    total_dim = nope_dim + rope_dim
+    BLOCK = 128  # Keep origin, works for smaller total_dim as well.
+    n_loc = loc.numel()
+    grid = (n_loc, triton.cdiv(total_dim, BLOCK))
```
Will `n_loc` become a large number? In which case there will be too many blocks in the grid.
This might be optimized in future PRs.

Fridge003 commented on python/sglang/srt/mem_cache/memory_pool.py:
```diff
@@ -1148,15 +1275,44 @@ def set_mla_kv_buffer(
+                cache_k_rope_fp4_sf,
+            )
+        else:
+            set_mla_kv_buffer_triton(
```
The `set_mla_kv_buffer_triton` kernel might also be moved to utils file

yicwang commented:
> Hi @JackChuang please fix the conflicts. Thanks

Both me and Jack are traveling attending conferences. We will be back next Monday and will be back working resolving this asap!

JackChuang commented:
> Good job, May I know if the sglang supports the FP8 mla kv cache or is there any plan for that? Thank you

Hi @zejunchen-zejun, Thank you! MLA's KV8 on SM100 has been supported. We ran DeepSeek-R1-0528-FP4-KV8_e4m3 in the experiment.

@

zejunchen-zejun commented:
> > Good job, May I know if the sglang supports the FP8 mla kv cache or is there any plan for that? Thank you
> 
> Hi @zejunchen-zejun, Thank you! MLA's KV8 on SM100 has been supported. We ran DeepSeek-R1-0528-FP4-KV8_e4m3 in the experiment.

Hi, @JackChuang 

Thank you for help! May I know if the support(FP8 mla kv cache) has been already merged in the main branch? Could you help specify the associated code ?
I checked the below code and found the mla set kv buffer cannot be passed scale argument, so I suspect it can only set the BF16 kv cache for now.
https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/mem_cache/memory_pool.py#L1333-L1339
```
    def set_mla_kv_buffer(
        self,
        layer: RadixAttention,
        loc: torch.Tensor,
        cache_k_nope: torch.Tensor,
        cache_k_rope: torch.Tensor,
    ):
```

CC: @qichu-yun

Thank you!

JackChuang commented:
> @JackChuang It shows there are still conflicts with latest main

@Fridge003  Oops.  You are right. It seems like v0.5.3 is also outdated. I will rebase to main then. Thanks, and sry for the inconvenience.

Fridge003 commented:
> Note: Turns out the current 'main' and 'v0.5.4.post1' branches cannot run. As a result, I’ve tested my code on v0.5.4, and it’s working.

Do you mean the v0.5.4.post1 cannot run on this PR, or is there other bugs?

Fridge003 commented on python/sglang/srt/layers/attention/trtllm_mla_backend.py:
```diff
@@ -360,19 +360,35 @@ def init_cuda_graph_state(
-            dtype=self.data_type,
-            device=self.device,
-        )
+        if self.data_type == getattr(torch, "float4_e2m1fn_x2", None) and _is_cuda:
```
One last comment

JackChuang committed:
Add is_float4_e2m1fn_x2() for repeated dtype checks

- Added `is_cuda()` and `is_float4_e2m1fn_x2()` in `sglang/srt/utils/torch_utils.py`
- Replaced inline checks in relevant modules

Signed-off-by: Ho-Ren (Jack) Chuang <horenchuang@bytedance.com>

Fridge003 commented:
@JackChuang Please fix CI bugs

Fridge003 commented:
https://github.com/sgl-project/sglang/actions/runs/18995002181/job/54283806641?pr=10078
